### PR TITLE
Abstracting out the layout rendering algorithms

### PR DESCRIPTION
**Pull request by marekyggdrasil**
This is my draft for abstracting out rendering algorithms for issue https://github.com/nicoespeon/gitgraph.js/issues/325

### CODE REVIEW

**Comment by marekyggdrasil** [comment]
@nicoespeon it's still a work-in-progress, please hold on with the review and certainly don't merge yet!

File: packages/gitgraph-core/src/layout-algorithms/gitamine.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { Commit } from "../commit";
+import FastPriorityQueue from "fastpriorityqueue";
```
Comment (andreasonny83) [inline packages/gitgraph-core/src/layout-algorithms/gitamine.ts:2]: can't find a reason for importing this

File: packages/gitgraph-core/package.json [hunk 2]
```diff
@@ -53,6 +53,9 @@
     "rollup": "^0.51.8",
     "uglify-es": "^3.3.9"
   },
+  "dependencies": {
+    "fastpriorityqueue": "^0.6.1"
```
Comment (andreasonny83) [inline packages/gitgraph-core/package.json:57]: can't find a valid reason for importing this library

**Review by andreasonny83 — CHANGES_REQUESTED** [review]

File: packages/gitgraph-core/package.json (hunk 2 above)
Comment (marekyggdrasil) [reply packages/gitgraph-core/package.json:57]: @pvigier do you mind if we drop the `fastpriorityqueue` library? I have no preference, I was basically following your gitamine design.

**Comment by marekyggdrasil** [comment]
@andreasonny83 thanks for taking your time to review this pull request, could you please also take a look at https://github.com/marekyggdrasil/gitgraph.js/pull/1 ? This is where @pvigier took over, I only prepared some foundations for him.

**Comment by Rain120** [comment]
### Error Result
The CI error result is the **nodejs** version must big than 10, you need to pull the master from this repo,  more info 👇👇👇

image

#### Master Code

image

### How to upgrade the new code

**Note**: Fix conflicts all of the code after fetched

```sh

# set remote upstream
git remote add upstream git@github.com:nicoespeon/gitgraph.js.git

# fetch it
git fetch upstream

```

More Info 👉👉👉  [Here](https://rain120.github.io/study-notes/#/notes/git-npm/git-tips?id=git-%e5%a6%82%e4%bd%95%e5%90%8c%e6%ad%a5fork-%e9%a1%b9%e7%9b%ae%e7%9a%84%e6%9c%80%e6%96%b0%e4%bb%a3%e7%a0%81)

**Comment by marekyggdrasil** [comment]
@Rain120 thanks for catching that. I just merged master branch from the upstream repository. The tests still fail but for totally different reasons, it's simply unfinished work.

**Comment by Rain120** [comment]
> @Rain120 thanks for catching that. I just merged master branch from the upstream repository. The tests still fail but for totally different reasons, it's simply unfinished work.

I saw the node version is 8.x of travis.yml configuate  at your repo.

**Pull request closed without merging**